### PR TITLE
ELU activation

### LINKED ISSUE #4364: ELU and SELU Operator
- Doc: http://pytorch.org/docs/master/nn.html#torch.nn.SELU
   - Activation in Paddle: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/activation_op.h

### PR DESCRIPTION
**Pull request by zhouxiao-coder**
Add ELU activation operator, resolve #4364 .
SELU activation is not added to the activation_op.{h cc cu} file, instead it can be trivially implemented with this ELU interface. Check #4364 for more details.

### CODE REVIEW

File: paddle/operators/activation_op.cc [hunk 1]
```diff
@@ -166,6 +166,23 @@ class SoftReluOpMaker : public framework::OpProtoAndCheckerMaker {
 };
 
 template <typename AttrType>
+class ELUOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  ELUOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "Input of ELU operator");
+    AddOutput("Y", "Output of ELU operator");
```
Comment (tonyyang-svail) [inline paddle/operators/activation_op.cc]: it could be change to `AddOutput("Y", "Output of ELU operator").NotInGradient();`, cause y is not used in `ELUGradKernel`.

Comment (zhouxiao-coder) [reply paddle/operators/activation_op.cc]: @tonyyang-svail Thanks for pointing that out, you are right.

However, after reexamining the math, I find there is a better way to rewrite the gradient formulation which does require using `y`. I'm working on it, and this PR will be updated soon.

File: paddle/operators/activation_op.cc [hunk 2]
```diff
@@ -166,6 +166,23 @@ class SoftReluOpMaker : public framework::OpProtoAndCheckerMaker {
 };
 
 template <typename AttrType>
+class ELUOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  ELUOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "Input of ELU operator");
```
Comment (lcy-seso) [inline paddle/operators/activation_op.cc]: Please follow our comment style https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/name_convention.md#opprotomaker-names

I think fc_op has a good comment style.

Comment (zhouxiao-coder) [reply paddle/operators/activation_op.cc]: I see. `fc_op` provides a nice reference indeed, and I'll try to follow that style.

File: python/paddle/v2/framework/tests/test_activation_op.py [hunk 3]
```diff
@@ -144,6 +144,24 @@ def test_check_grad(self):
...
 
+class TestELU(OpTest):
+    def setUp(self):
+        self.op_type = "elu"
+        x = np.random.uniform(-3, 3, [4, 4]).astype("float32")
+        alpha = 1.
+        self.inputs = {'X': x}
+        self.attrs = {'alpha': alpha}
+        self.outputs = {
+            'Y': np.maximum(0, x) + np.minimum(0, alpha * (np.exp(x) - 1))
```
[the remark this answers is not in the document]
Comment (zhouxiao-coder) [reply python/paddle/v2/framework/tests/test_activation_op.py]: Thanks for the reference, I'm a little bit confused by the line `x[np.abs(x) < 0.005] = 0.02` before and #4120 explains the motivation.

However, ELU has a quite "smooth" negative part, so this modification may be unnecessary here. In fact, `elu(x=-0.005)` gives -0.0049875208073176802, which makes the relative numeric gradient error less than 2e-5 and is small enough to pass normal gradient check.
Here is the code I used to experiment.
```
import matplotlib.pyplot as plt
import numpy as np
def elu(x):
    return np.maximum(0, x) + np.minimum(0, 1. * (np.exp(x) - 1))
x = np.linspace(-0.5, 0.5, num=100)
y = elu(x)
plt.plot(x,y)
```
figure_1

[the remark this answers is not in the document]
Comment (zhouxiao-coder) [reply python/paddle/v2/framework/tests/test_activation_op.py]: But on the other hand, I can make another PR if you think we should make this modification `x[np.abs(x) < 0.005] = 0.02` a common practice in unit test.

[the remark this answers is not in the document]
Comment (QiJune) [reply python/paddle/v2/framework/tests/test_activation_op.py]: Yes, elu is actually very smooth. 0 is a non-differentiable point in elu. We'd better to filter it in out test data.
This line
```
x[np.abs(x) < 0.005] = 0.02
```
will filter these potential points and set them to 0.02.(0.02-0.005 = 0.015;0.02+0.005=0.025, both of 0.015 and 0.025 are larger than 0)

[the remark this answers is not in the document]
Comment (zhouxiao-coder) [reply python/paddle/v2/framework/tests/test_activation_op.py]: As long as we are talking about the standard `elu`, i.e. alpha=1, then it is differentiable at point 0 since both left derivate and right derivate at 0 equal to 1. The "piecewise" form of the function only makes it non-differentiable on the derivative function at 0, i.e. the second order derivative.

File: paddle/operators/activation_op.cc (hunk 2 above)
Comment (zhouxiao-coder) [reply paddle/operators/activation_op.cc]: Done. I didn't squash all my commit because they get interleaved with other's commit. Please check commit 4436ba0 for the new comments. Sorry for the inconvenience

File: python/paddle/v2/framework/tests/test_activation_op.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhouxiao-coder) [reply python/paddle/v2/framework/tests/test_activation_op.py]: @QiJune I didn't add the filtering code to the new commit, since as we discussed it is not necessary. However, I add a comment to explain the reason.

File: paddle/operators/activation_op.cc (hunk 1 above)
Comment (zhouxiao-coder) [reply paddle/operators/activation_op.cc]: Done. 
Now the gradient of negative part is computed as follow:
'''
dy * (y + alpha) * (x < static_cast<T>(0)).template cast<T>();
'''
So now value of `y` is used in the gradient calculation.
